pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/merge_rollup/MergeRollupTaskGenerator.java

```diff
@@ -0,0 +1,461 @@
+import org.apache.pinot.spi.utils.TimeUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * A {@link PinotTaskGenerator} implementation for generating tasks of type {@link MergeRollupTask}
+ *
+ * TODO: Add the support for realtime table
+ *
+ * Steps:
+ *
+ *  - Pre-select segments:
+ *    - Fetch all segments, select segments based on segment lineage (removing segmentsFrom for COMPLETED lineage entry and
+ *      segmentsTo for IN_PROGRESS lineage entry)
+ *    - Sort segments based on startTime and endTime in ascending order
+ *
+ *  For each mergeLevel (from lowest to highest, e.g. Hourly -> Daily -> Monthly -> Yearly):
+ *    - Skip scheduling if there's incomplete task for the mergeLevel
+ *    - Calculate merge/rollup window:
+ *      - Read watermarkMs from the {@link MergeRollupTaskMetadata} ZNode
+ *        found at MINION_TASK_METADATA/MergeRollupTask/<tableNameWithType>
+ *        In case of cold-start, no ZNode will exist.
+ *        A new ZNode will be created, with watermarkMs as the smallest time found in all segments truncated to the
+ *        closest bucket start time.
+ *      - The execution window for the task is calculated as,
+ *        windowStartMs = watermarkMs, windowEndMs = windowStartMs + bucketTimeMs
+ *      - Skip scheduling if the window is invalid:
+ *        - If the execution window is not older than bufferTimeMs, no task will be generated
+ *        - The windowEndMs of higher merge level should be less or equal to the waterMarkMs of lower level
+ *      - Bump up target window and watermark if needed.
+ *        - If there's no unmerged segments (by checking segment zk metadata {mergeRollupTask.mergeLevel: level}) for current window,
+ *          keep bumping up the watermark and target window until unmerged segments are found. Else skip the scheduling.
+ *    - Select all segments for the target window
+ *    - Create tasks (per partition for partitioned table) based on maxNumRecordsPerTask
+ */
+@TaskGenerator
+public class MergeRollupTaskGenerator implements PinotTaskGenerator {
+  private static final int DEFAULT_MAX_NUM_RECORDS_PER_TASK = 50_000_000;
+  private static final String REFRESH = "REFRESH";
+  private static final Logger LOGGER = LoggerFactory.getLogger(MergeRollupTaskGenerator.class);
+
+  private ClusterInfoAccessor _clusterInfoAccessor;
+
+  @Override
+  public void init(ClusterInfoAccessor clusterInfoAccessor) {
+    _clusterInfoAccessor = clusterInfoAccessor;
+  }
+
+  @Override
+  public String getTaskType() {
+    return MergeRollupTask.TASK_TYPE;
+  }
+
+  @Override
+  public List<PinotTaskConfig> generateTasks(List<TableConfig> tableConfigs) {
+    String taskType = MergeRollupTask.TASK_TYPE;
+    List<PinotTaskConfig> pinotTaskConfigs = new ArrayList<>();
+
+    for (TableConfig tableConfig : tableConfigs) {
+      String offlineTableName = tableConfig.getTableName();
+
+      if (tableConfig.getTableType() != TableType.OFFLINE) {
```

Comment:
One way to reduce the number of lines of code for this function is to use sub-functions.  For instance, we can put line 113-130 into a separate function and call something like `validate(TableConfig tableConfig)` and put all the validation-related logic into the function. You can decide whether we need this.

Reply:
Moved to a helper function.